File: packages/clash/src/engine-ts/narrow.ts
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import type { AABB, ClashDistanceKind, ClashElement, ClashRule, ClashStatus, Vec3 } from '../types.js';
import { aabbContains, boundsOfPoints, center, inflate, overlapBounds, signedGap } from '../math/aabb.js';
import { centroid, mid } from '../math/vec3.js';
import { triTriIntersect } from '../math/triangle-intersect.js';
import { triTriDistance } from '../math/triangle-distance.js';
import type { TriMesh } from './tri-mesh.js';
import { isThroughPenetration, obbPenetrationDepth } from './obb.js';

/**
 * Exact box-box penetration depth when BOTH meshes are (within tolerance)
 * rectangular boxes, else `null`. This is the only source of a `'mesh'`
 * label for a distance that used to come from `TriMesh.maxPenetrationInto` —
 * a nearest-crossing-vertex sampling probe that converges to 0 under
 * retessellation instead of to the true depth (see `obb.ts`, `obb.test.ts`).
 *
 * Declines (returns `null`, same as "not both boxes") for a THROUGH-
 * PENETRATION pair — a thin member piercing clean through the other, e.g. a
 * duct through a wall. There, the minimum-translation-distance `obb.ts`
 * computes is dominated by the piercing member's own extent along the
 * shared axis, not by the material actually crossed (review: #2536,
 * reproduced a 5.5x inflation on exactly this shape). Falling back to the
 * caller's AABB estimate for that shape matches what `main` did before the
 * box-exact metric existed, and is honest about not being a measurement.
 */
function boxMeasuredDepth(small: TriMesh, large: TriMesh): number | null {
  const oa = small.getObb();
  const ob = large.getObb();
  if (!oa || !ob) return null;
  if (isThroughPenetration(oa, ob)) return null;
  return obbPenetrationDepth(oa, ob);
}

export interface NarrowResult {
  status: ClashStatus;
  distance: number;
  /**
   * Whether `distance` was measured on the meshes or estimated from the AABBs.
   * Set on every result, so a caller never has to guess which of the two very
   * different quantities it is holding.
   */
  distanceKind: ClashDistanceKind;
  point: Vec3;
  bounds: AABB;
}

/**
 * Narrow-phase test for one candidate element pair.
 *
 * Gathers candidate triangle pairs through the per-element triangle BVHs (work
 * stays proportional to actual overlap — no decimation), then:
 * - a genuine non-coplanar triangle crossing ⇒ `hard` (pipe through beam, etc.);
 * - surfaces that merely coincide/touch (no crossing) are disambiguated by AABB
 *   penetration, so a deep axis-aligned overlap — whose surface intersections
 *   are all coplanar — is still classified `hard`, while face contact is `touch`;
 * - otherwise the exact triangle distance drives `clearance` / `touch` / no clash.
 *
 * Returns `null` for no clash.
 */
export function testPair(
  elA: ClashElement,
  triA: TriMesh,
  elB: ClashElement,
  triB: TriMesh,
  rule: ClashRule,
  tolerance: number,
): NarrowResult | null {
  const margin = Math.max(tolerance, rule.clearance ?? 0);

  // Iterate the smaller mesh, querying the larger one's BVH.
  const aSmaller = triA.count <= triB.count;
  const small = aSmaller ? triA : triB;
  const large = aSmaller ? triB : triA;

  let intersects = false;
  let contactSumX = 0;
  let contactSumY = 0;
  let contactSumZ = 0;
  let contactN = 0;
  // Tight contact AABB: min/max of the per-pair contact points (the crossing
  // representatives), so a hard verdict reports the local contact region rather
  // than the whole-element AABB overlap (#1362 / #1402).
  const cMin: Vec3 = [Infinity, Infinity, Infinity];
  const cMax: Vec3 = [-Infinity, -Infinity, -Infinity];
  // Near-contact AABB for coplanar/flush overlaps (no triangle crossing): the
  // local region where surfaces actually touch, so the hard box is the contact
  // patch (e.g. a wall corner) rather than the whole-element AABB intersection,
  // which for angled members spans nearly the full member length (#1362/#1402).
  const ncMin: Vec3 = [Infinity, Infinity, Infinity];
  const ncMax: Vec3 = [-Infinity, -Infinity, -Infinity];
  let ncN = 0;
  let minDist = Infinity;
  let closestA: Vec3 = elA.bounds.min as Vec3;
  let closestB: Vec3 = elB.bounds.min as Vec3;

  for (let ts = 0; ts < small.count; ts += 1) {
    const sb = small.triBounds(ts);
    const hits = large.queryTris(inflate(sb, margin));
    if (hits.length === 0) continue;
    const [s0, s1, s2] = small.tri(ts);
    for (const tl of hits) {
      const [l0, l1, l2] = large.tri(tl);
      if (triTriIntersect(s0, s1, s2, l0, l1, l2)) {
        intersects = true;
        const c = mid(centroid(s0, s1, s2), centroid(l0, l1, l2));
        contactSumX += c[0];
        contactSumY += c[1];
        contactSumZ += c[2];
        contactN += 1;
        for (let i = 0; i < 3; i += 1) {
          if (c[i] < cMin[i]) cMin[i] = c[i];
          if (c[i] > cMax[i]) cMax[i] = c[i];
        }
      } else {
        // Not a crossing: measure the gap (drives clearance/touch) and, when the
        // pair is touching (within tolerance), accumulate it into the contact
        // region. We do this even after a crossing was found, because coincident
        // faces of flush members register as touches (not crossings) and carry
        // most of the real contact area.
        const d = triTriDistance(s0, s1, s2, l0, l1, l2);
        if (d.dist < minDist) {
          minDist = d.dist;
          closestA = d.pA;
          closestB = d.pB;
        }
        if (d.dist <= tolerance) {
          const cp = mid(d.pA, d.pB);
          ncN += 1;
          for (let i = 0; i < 3; i += 1) {
            if (cp[i] < ncMin[i]) ncMin[i] = cp[i];
            if (cp[i] > ncMax[i]) ncMax[i] = cp[i];
          }
        }
      }
    }
  }

  const overlap = overlapBounds(elA.bounds, elB.bounds);

  // Tight contact region: the union of the genuine triangle crossings (cMin/cMax)
  // and the coplanar/flush touching pairs within tolerance (ncMin/ncMax), clamped
  // to the element overlap. Crossings alone miss coincident faces (which register
  // as touches, not crossings) so flush members reported only a partial, mis-
  // placed patch; near-contacts alone miss angled crossings. Falls back to the
  // overlap when neither was captured (#1362 / #1402).
  const tMin: Vec3 = [Infinity, Infinity, Infinity];
  const tMax: Vec3 = [-Infinity, -Infinity, -Infinity];
  let tN = 0;
  if (contactN > 0) { for (let i = 0; i < 3; i += 1) { if (cMin[i] < tMin[i]) tMin[i] = cMin[i]; if (cMax[i] > tMax[i]) tMax[i] = cMax[i]; } tN += 1; }
  if (ncN > 0) { for (let i = 0; i < 3; i += 1) { if (ncMin[i] < tMin[i]) tMin[i] = ncMin[i]; if (ncMax[i] > tMax[i]) tMax[i] = ncMax[i]; } tN += 1; }
  // Clamp the contact AABB to the element overlap per-axis. (overlapBounds would
  // degenerate a disjoint axis to a midpoint that can land OUTSIDE the overlap,
  // breaking the "clamped to overlap" contract for the box.)
  let contactBounds = overlap;
  if (tN > 0) {
    const cMinClamped: Vec3 = [0, 0, 0];
    const cMaxClamped: Vec3 = [0, 0, 0];
    for (let i = 0; i < 3; i += 1) {
      cMinClamped[i] = Math.min(Math.max(tMin[i], overlap.min[i]), overlap.max[i]);
      cMaxClamped[i] = Math.min(Math.max(tMax[i], overlap.min[i]), overlap.max[i]);
    }
    contactBounds = { min: cMinClamped, max: cMaxClamped };
  }

  if (intersects) {
    const point: Vec3 = contactN > 0
      ? [contactSumX / contactN, contactSumY / contactN, contactSumZ / contactN]
      : center(overlap);
    // Exact box-box penetration depth when both elements are rectangular
    // boxes (see `boxMeasuredDepth`); otherwise fall back to the AABB overlap,
    // i.e. the smallest overlapping box dimension. That fallback is an
    // estimate, not a measured depth — for a non-box shape it can report a
    // dimension of one of the elements rather than how far they interpenetrate.
    const boxDepth = boxMeasuredDepth(small, large);
    const measured = boxDepth != null;
    const penetration = measured
      ? boxDepth
      : Math.max(0, -signedGap(elA.bounds, elB.bounds));
    return {
      status: 'hard',
      distance: -penetration,
      distanceKind: measured ? 'mesh' : 'estimate',
      point,
      bounds: contactBounds,
    };
  }

  // Fully-enclosed solid: no surface crossing, but one element's AABB is wholly
  // inside the other's, so it may be buried (e.g. equipment inside a slab). With
  // no surface crossing the inner solid is entirely inside OR entirely outside
  // the other, so ray-casting ONE representative vertex of the contained mesh
  // against the other solid decides it — and ray casting (not an AABB test)
  // correctly returns "outside" when the inner sits in a concave notch.
  // Test B-contains-A first, then A-contains-B, so the inner pick is
  // deterministic (and identical to the Rust kernel) on equal AABBs.
  // Either way there is no surface crossing. When both elements are boxes the
  // exact box-box depth is available (see `boxMeasuredDepth`) and is reported
  // as measured; otherwise the AABB gap is an estimate, not a measured depth.
  if (aabbContains(elB.bounds, elA.bounds)) {
    if (triA.count > 0 && triB.containsPoint(triA.tri(0)[0])) {
      const boxDepth = boxMeasuredDepth(small, large);
      return boxDepth != null
        ? { status: 'hard', distance: -boxDepth, distanceKind: 'mesh', point: center(overlap), bounds: overlap }
        : { status: 'hard', distance: signedGap(elA.bounds, elB.bounds), distanceKind: 'estimate', point: center(overlap), bounds: overlap };
    }
  } else if (aabbContains(elA.bounds, elB.bounds)) {
    if (triB.count > 0 && triA.containsPoint(triB.tri(0)[0])) {
      const boxDepth = boxMeasuredDepth(small, large);
      return boxDepth != null
        ? { status: 'hard', distance: -boxDepth, distanceKind: 'mesh', point: center(overlap), bounds: overlap }
        : { status: 'hard', distance: signedGap(elA.bounds, elB.bounds), distanceKind: 'estimate', point: center(overlap), bounds: overlap };
    }
  }

  if (minDist === Infinity) {
    // Broad-phase candidate with no triangle-level proximity — not a clash.
    return null;
  }

  // Surfaces coincide/touch with no genuine crossing, but the AABBs penetrate
  // beyond tolerance (e.g. axis-aligned boxes, whose surface intersections are
  // all coplanar). AABB penetration ALONE is not enough: two skewed/abutting
  // members that merely share a face have overlapping AABBs yet no shared volume,
  // and the old proxy promoted that touch to a false hard clash (#1362). Confirm
  // a real shared volume first by probing for an interior point inside BOTH
  // solids. Two probes are needed: the vertex-centroid midpoint sits inside a
  // skewed straddling overlap, while the AABB-overlap centre covers an
  // unequal-length aligned overlap (whose centroid midpoint can fall outside the
  // shorter member). A bare face touch has no interior point common to both, so
  // neither probe qualifies. Accept the pair if EITHER probe is inside both.
  if (minDist <= tolerance) {
    const gap = signedGap(elA.bounds, elB.bounds);
    if (gap < -tolerance) {
      const probeCentroid = mid(triA.vertexCentroid(), triB.vertexCentroid());
      const probeOverlap = center(overlap);
      if (
        (triA.containsPoint(probeCentroid) && triB.containsPoint(probeCentroid)) ||
        (triA.containsPoint(probeOverlap) && triB.containsPoint(probeOverlap))
      ) {
        // Report the tight contact region (the touching patch where the surfaces
        // actually coincide), clamped to the element overlap — not the whole-
        // element AABB intersection, which for angled members spans nearly the
        // full member length and sits away from the real contact (#1362/#1402).
        // When both elements are boxes the exact box-box depth is available
        // (see `boxMeasuredDepth`) and is reported as measured, not estimated.
        const boxDepth = boxMeasuredDepth(small, large);
        return {
          status: 'hard',
          distance: boxDepth != null ? -boxDepth : gap,
          distanceKind: boxDepth != null ? 'mesh' : 'estimate',
          point: mid(closestA, closestB),
          bounds: contactBounds,
        };
      }
      // Only a face touch (no shared volume): fall through to the touch handling
      // below, which suppresses it unless reportTouch is set.
    }
  }

  // Clearance rule: ANY gap within the required clearance is a violation —
  // including sub-tolerance, nearly-touching gaps, which are the MOST severe and
  // must not be swallowed by the touch band below.
  if (rule.mode === 'clearance' && rule.clearance != null && minDist <= rule.clearance) {
    return {
      status: 'clearance',
      distance: minDist,
      // `minDist` is an exact triangle-to-triangle distance.
      distanceKind: 'mesh',
      point: mid(closestA, closestB),
      bounds: boundsOfPoints(closestA, closestB),
    };
  }

  // Otherwise only bare contact within tolerance remains; suppressed unless the
  // rule opts in. `<=` so an exact touch at tolerance 0 is still caught.
  if (minDist <= tolerance) {
    if (!rule.reportTouch) return null;
    return {
      status: 'touch',
      distance: minDist,
      distanceKind: 'mesh',
      point: mid(closestA, closestB),
      bounds: boundsOfPoints(closestA, closestB),
    };
  }

  return null;
}

File: rust/clash/src/narrow.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

//! Narrow-phase classification for one candidate element pair.
//!
//! Faithful port of `packages/clash/src/engine-ts/narrow.ts`. The control flow,
//! comparisons, and result construction match the TS reference bit-for-bit in
//! logic so this kernel and the TS engine agree on classification.

use crate::aabb::{aabb_contains, bounds_of_points, overlap_bounds, signed_gap, Aabb};
use crate::triangle::{tri_tri_distance, tri_tri_intersect};
use crate::tri_mesh::TriMesh;
use crate::vec3::{centroid, mid, Vec3};

/// Clash classification. Discriminants match the public ABI (`Hard = 0`, etc.).
#[derive(Clone, Copy, PartialEq, Debug)]
pub enum ClashStatus {
    Hard = 0,
    Clearance = 1,
    Touch = 2,
}

/// The narrow-phase outcome for one element pair.
pub struct NarrowResult {
    pub status: ClashStatus,
    pub distance: f64,
    pub point: Vec3,
    pub bounds: Aabb,
}

/// f32-ULP scale factor for a "worst-case" single-precision coordinate: for a
/// value with magnitude in `[2, 4)` the true float32 ULP is `2^-22`, and for
/// larger magnitudes the ULP only grows. Same `2^-22` term (and reasoning) as
/// `near_band_from_extent` in `rust/geometry/src/kernel/mesh_bridge.rs` — see
/// that function's doc for the derivation; kept here rather than shared
/// because the two crates serve different callers.
const F32_ULP_SCALE: f64 = 1.0 / 4_194_304.0; // 2^-22

/// Penetration-depth floor below which a computed overlap cannot be
/// distinguished from float32 rounding noise, scaled to the pair's own
/// coordinate magnitude (not a fixed constant — infra models sit far from the
/// origin, where a fixed epsilon would be far too tight, and small models sit
/// near it, where a fixed epsilon would be far too loose).
///
/// `tri_mesh.rs` ingests geometry from f32 buffers and stores/queries it in
/// f64, so f64 arithmetic cannot recover precision the source data never
/// had: two surfaces authored to be flush round to adjacent f32 values, and
/// the resulting "penetration" is bit-noise at the ULP of whichever operand
/// coordinate is largest, not a measured overlap. Extent is the max abs
/// coordinate over both elements' AABBs (matching `near_band_from_extent`'s
/// use of the actual compared coordinates), floored at 1.0 so a model near
/// the origin still gets the single-unit ULP, not zero.
///
/// The floor grows linearly with the pair's distance from the origin — that
/// is the point, since f32 precision itself degrades the same way. The
/// consequence: on a georeferenced model whose elements sit at real map
/// coordinates (hundreds of kilometres out, which real IFC files do), the
/// floor reaches decimetre scale, and a genuine clash below that threshold
/// reclassifies as `Touch`. That is not a bug in this function — at those
/// magnitudes f32 genuinely cannot represent a finer distinction, so the
/// "penetration" is not reliably measurable either way — but it means the
/// floor tracks a limitation of the source data, not of clash detection.
/// The fix for a model in that position is ingesting geometry closer to the
/// origin (or in f64), not lowering this floor.
fn precision_floor(aabb_a: &Aabb, aabb_b: &Aabb) -> f64 {
    let mut extent = 1.0f64;
    for b in [aabb_a, aabb_b] {
        for v in [&b.min, &b.max] {
            for &c in v {
                let a = c.abs();
                if a > extent {
                    extent = a;
                }
            }
        }
    }
    extent * F32_ULP_SCALE
}

/// Run the narrow phase for a candidate element pair.
///
/// `mode`: `0` = hard, `1` = clearance. `tolerance` and `clearance` carry the
/// rule parameters; `report_touch` toggles face-contact reporting. Returns
/// `None` when the pair is not a clash.
#[allow(clippy::too_many_arguments)]
pub fn test_pair(
    aabb_a: &Aabb,
    tri_a: &TriMesh,
    aabb_b: &Aabb,
    tri_b: &TriMesh,
    mode: u8,
    tolerance: f64,
    clearance: f64,
    report_touch: bool,
) -> Option<NarrowResult> {
    let is_clearance = mode == 1;
    let margin = tolerance.max(if is_clearance { clearance } else { 0.0 });

    // Iterate the smaller mesh, querying the larger one's BVH.
    let a_smaller = tri_a.count <= tri_b.count;
    let (small, large) = if a_smaller {
        (tri_a, tri_b)
    } else {
        (tri_b, tri_a)
    };

    // One AABB containing the other flags the contained-contact case (#1866):
    // for such pairs the AABB signed gap measures how deep the small BOX sits in
    // the big one (its own extent), not how far the MESHES interpenetrate, so
    // collect the crossing triangles for a mesh-level depth measurement instead.
    let contained = aabb_contains(aabb_b, aabb_a) || aabb_contains(aabb_a, aabb_b);
    let mut cross_small: Vec<bool> = if contained { vec![false; small.count] } else { Vec::new() };
    let mut cross_large: Vec<bool> = if contained { vec![false; large.count] } else { Vec::new() };

    let mut intersects = false;
    let mut contact_sum: [f64; 3] = [0.0, 0.0, 0.0];
    let mut contact_n: u32 = 0;
    // Tight contact AABB: min/max of the per-pair contact points (the crossing
    // representatives), so a hard verdict reports the local contact region rather
    // than the whole-element AABB overlap (#1362 / #1402).
    let mut c_min: Vec3 = [f64::INFINITY; 3];
    let mut c_max: Vec3 = [f64::NEG_INFINITY; 3];
    // Near-contact AABB for coplanar/flush overlaps (no triangle crossing): the
    // local touching region, so the hard box is the contact patch (e.g. a wall
    // corner) not the whole-element AABB intersection, which for angled members
    // spans nearly the full member length (#1362 / #1402).
    let mut nc_min: Vec3 = [f64::INFINITY; 3];
    let mut nc_max: Vec3 = [f64::NEG_INFINITY; 3];
    let mut nc_n: u32 = 0;
    let mut min_dist = f64::INFINITY;
    let mut closest_a: Vec3 = aabb_a.min;
    let mut closest_b: Vec3 = aabb_b.min;

    for ts in 0..small.count {
        let sb = small.tri_bounds(ts);
        let hits = large.query_tris(&sb.inflate(margin));
        if hits.is_empty() {
            continue;
        }
        let [s0, s1, s2] = small.tri(ts);
        for tl in hits {
            let [l0, l1, l2] = large.tri(tl as usize);
            if tri_tri_intersect(s0, s1, s2, l0, l1, l2) {
                intersects = true;
                // Flag the crossing pair for the contained-case depth
                // measurement; the flag vecs are empty (`get_mut` = None)
                // when the pair is not contained.
                if let Some(flag) = cross_small.get_mut(ts) {
                    *flag = true;
                }
                if let Some(flag) = cross_large.get_mut(tl as usize) {
                    *flag = true;
                }
                let c = mid(centroid(s0, s1, s2), centroid(l0, l1, l2));
                contact_sum[0] += c[0];
                contact_sum[1] += c[1];
                contact_sum[2] += c[2];
                contact_n += 1;
                for i in 0..3 {
                    if c[i] < c_min[i] {
                        c_min[i] = c[i];
                    }
                    if c[i] > c_max[i] {
                        c_max[i] = c[i];
                    }
                }
            } else {
                // Not a crossing: measure the gap (drives clearance/touch) and,
                // when touching (within tolerance), accumulate the pair into the
                // contact region. Done even after a crossing is found, since
                // coincident faces of flush members register as touches (not
                // crossings) yet carry most of the real contact area.
                let (dist, p_a, p_b) = tri_tri_distance(s0, s1, s2, l0, l1, l2);
                if dist < min_dist {
                    min_dist = dist;
                    closest_a = p_a;
                    closest_b = p_b;
                }
                if dist <= tolerance {
                    let cp = mid(p_a, p_b);
                    nc_n += 1;
                    for i in 0..3 {
                        if cp[i] < nc_min[i] {
                            nc_min[i] = cp[i];
                        }
                        if cp[i] > nc_max[i] {
                            nc_max[i] = cp[i];
                        }
                    }
                }
            }
        }
    }

    let overlap = overlap_bounds(aabb_a, aabb_b);

    // Tight contact region: the union of the genuine triangle crossings
    // (c_min/c_max) and the coplanar/flush touching pairs within tolerance
    // (nc_min/nc_max), clamped to the element overlap. Crossings alone miss
    // coincident faces (which register as touches, not crossings) so flush members
    // reported only a partial, mis-placed patch; near-contacts alone miss angled
    // crossings. Falls back to the overlap when neither was captured (#1362/#1402).
    let mut t_min: Vec3 = [f64::INFINITY; 3];
    let mut t_max: Vec3 = [f64::NEG_INFINITY; 3];
    let mut t_n: u32 = 0;
    if contact_n > 0 {
        for i in 0..3 {
            if c_min[i] < t_min[i] {
                t_min[i] = c_min[i];
            }
            if c_max[i] > t_max[i] {
                t_max[i] = c_max[i];
            }
        }
        t_n += 1;
    }
    if nc_n > 0 {
        for i in 0..3 {
            if nc_min[i] < t_min[i] {
                t_min[i] = nc_min[i];
            }
            if nc_max[i] > t_max[i] {
                t_max[i] = nc_max[i];
            }
        }
        t_n += 1;
    }
    let contact_bounds = if t_n > 0 {
        // Clamp the contact AABB to the element overlap per-axis. (overlap_bounds
        // would degenerate a disjoint axis to a midpoint that can land OUTSIDE the
        // overlap, breaking the "clamped to overlap" contract for the box.)
        let mut min: Vec3 = [0.0; 3];
        let mut max: Vec3 = [0.0; 3];
        for i in 0..3 {
            min[i] = t_min[i].max(overlap.min[i]).min(overlap.max[i]);
            max[i] = t_max[i].max(overlap.min[i]).min(overlap.max[i]);
        }
        Aabb::new(min, max)
    } else {
        overlap
    };

    if intersects {
        let point: Vec3 = if contact_n > 0 {
            let n = contact_n as f64;
            [contact_sum[0] / n, contact_sum[1] / n, contact_sum[2] / n]
        } else {
            overlap.center()
        };
        // Penetration estimate from the AABB overlap...
        let mut penetration = (-signed_gap(aabb_a, aabb_b)).max(0.0);
        // ...EXCEPT for a contained pair (#1866): there the AABB overlap equals
        // the small element's own extent, wildly overstating depth for designed
        // face contacts (e.g. opening fills inset in their host). Measure the
        // real mesh-level depth instead: the deepest crossing-triangle vertex of
        // either mesh inside the other solid. Falls back to the AABB estimate
        // when no such vertex lies inside (thin member piercing straight
        // through).
        if contained {
            let mesh_depth = small
                .max_penetration_into(large, &cross_small)
                .max(large.max_penetration_into(small, &cross_large));
            if mesh_depth > 0.0 {
                penetration = mesh_depth;
            }
        }
        // A genuine triangle crossing was found, but the measured depth is at
        // or below the f32 precision floor for this pair's coordinate scale:
        // it cannot be distinguished from rounding noise (see
        // `precision_floor`), so it is not a measured overlap. Reclassify as
        // `Touch` rather than `Hard` — the crossing still means the surfaces
        // are in contact, which is real information (this codebase already
        // distinguishes touching from overlapping, e.g. the viewer's
        // `clashHideTouching` toggle) — rather than silently dropping the
        // pair or reporting a fabricated depth.
        if penetration <= precision_floor(aabb_a, aabb_b) {
            if !report_touch {
                return None;
            }
            return Some(NarrowResult {
                status: ClashStatus::Touch,
                distance: 0.0,
                point,
                bounds: contact_bounds,
            });
        }
        return Some(NarrowResult {
            status: ClashStatus::Hard,
            distance: -penetration,
            point,
            bounds: contact_bounds,
        });
    }

    // Fully-enclosed solid: no surface crossing, but one element's AABB is wholly
    // inside the other's, so it may be buried. With no surface crossing the inner
    // solid is entirely inside OR entirely outside the other, so ray-casting ONE
    // representative vertex of the contained mesh decides it — and ray casting
    // (not an AABB test) correctly returns "outside" for a concave-notch case.
    // Test B-contains-A first, then A-contains-B, so the inner pick is
    // deterministic (and identical to the TS kernel) on equal AABBs.
    let enclosed = if aabb_contains(aabb_b, aabb_a) {
        tri_a.count > 0 && tri_b.contains_point(tri_a.tri(0)[0])
    } else if aabb_contains(aabb_a, aabb_b) {
        tri_b.count > 0 && tri_a.contains_point(tri_b.tri(0)[0])
    } else {
        false
    };
    if enclosed {
        return Some(NarrowResult {
            status: ClashStatus::Hard,
            distance: signed_gap(aabb_a, aabb_b),
            point: overlap.center(),
            bounds: overlap,
        });
    }

    if min_dist == f64::INFINITY {
        // Broad-phase candidate with no triangle-level proximity — not a clash.
        return None;
    }

    // Surfaces coincide/touch with no genuine crossing, but the AABBs penetrate
    // beyond tolerance (coplanar surfaces, e.g. axis-aligned boxes). AABB
    // penetration ALONE is not enough: two skewed/abutting members that merely
    // share a face have overlapping AABBs yet no shared volume, and the old proxy
    // promoted that touch to a false hard clash (#1362). Confirm a real shared
    // volume first by probing for an interior point inside BOTH solids. Two probes
    // are needed: the vertex-centroid midpoint sits inside a skewed straddling
    // overlap, while the AABB-overlap centre covers an unequal-length aligned
    // overlap (whose centroid midpoint can fall outside the shorter member). A
    // bare face touch has no interior point common to both, so neither probe
    // qualifies. Accept the pair if EITHER probe is inside both.
    if min_dist <= tolerance {
        let gap = signed_gap(aabb_a, aabb_b);
        if gap < -tolerance {
            let probe_centroid = mid(tri_a.vertex_centroid(), tri_b.vertex_centroid());
            let probe_overlap = overlap.center();
            if (tri_a.contains_point(probe_centroid) && tri_b.contains_point(probe_centroid))
                || (tri_a.contains_point(probe_overlap) && tri_b.contains_point(probe_overlap))
            {
                // Report the tight contact region (the touching patch where the
                // surfaces actually coincide), clamped to the element overlap — not
                // the whole-element AABB intersection, which for angled members
                // spans nearly the full member length and sits away from the real
                // contact (#1362/#1402).
                return Some(NarrowResult {
                    status: ClashStatus::Hard,
                    distance: gap,
                    point: mid(closest_a, closest_b),
                    bounds: contact_bounds,
                });
            }
            // Only a face touch (no shared volume): fall through to the touch
            // handling below, which suppresses it unless report_touch is set.
        }
    }

    // Clearance rule: ANY gap within the required clearance is a violation,
    // including sub-tolerance, nearly-touching gaps (the most severe). These
    // must not be swallowed by the touch band below.
    if is_clearance && min_dist <= clearance {
        return Some(NarrowResult {
            status: ClashStatus::Clearance,
            distance: min_dist,
            point: mid(closest_a, closest_b),
            bounds: bounds_of_points(closest_a, closest_b),
        });
    }

    // Otherwise only bare contact within tolerance remains; suppressed unless the
    // rule opts in. `<=` so an exact touch at tolerance 0 is still caught.
    if min_dist <= tolerance {
        if !report_touch {
            return None;
        }
        return Some(NarrowResult {
            status: ClashStatus::Touch,
            distance: min_dist,
            point: mid(closest_a, closest_b),
            bounds: bounds_of_points(closest_a, closest_b),
        });
    }

    None
}
